### ISSUE TITLE
Unable to restore Bioconductor packages 

### ISSUE THREAD

**Issue by PieGes** [rstudio/renv#244]
When trying to restore Bioconductor packages, `renv::restore` tries to download packages related to the latest Bioc version (3.10 released on October 30) even if the snapshot was done on Bioc version 3.9. I guess the version should be specified when setting BiocManager repositories in `renv_bioconductor_repos`.

Here is the lockfile

```
{
  "R": {
    "Version": "3.6.0",
    "Repositories": [
      {
        "Name": "BioCsoft",
        "URL": "https://bioconductor.org/packages/3.9/bioc"
      },
      {
        "Name": "BioCann",
        "URL": "https://bioconductor.org/packages/3.9/data/annotation"
      },
      {
        "Name": "BioCexp",
        "URL": "https://bioconductor.org/packages/3.9/data/experiment"
      },
      {
        "Name": "BioCworkflows",
        "URL": "https://bioconductor.org/packages/3.9/workflows"
      },
      {
        "Name": "CRAN",
        "URL": "https://cloud.r-project.org"
      }
    ]
  },
  "Packages": {
    "Biobase": {
      "Package": "Biobase",
      "Version": "2.44.0",
      "Source": "Bioconductor",
      "Hash": "4d603778097e0917bcd309eff2f314f9"
    },
    "BiocGenerics": {
      "Package": "BiocGenerics",
      "Version": "0.30.0",
      "Source": "Bioconductor",
      "Hash": "0e4a1c136ce54159dbc4743379861d92"
    },
    "BiocManager": {
      "Package": "BiocManager",
      "Version": "1.30.9",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "ae39a770c7be4816960bde0c25626b20"
    },
    "KernSmooth": {
      "Package": "KernSmooth",
      "Version": "2.23-15",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "1abaf80e4ce51c982446637b6cf236bb"
    },
    "MASS": {
      "Package": "MASS",
      "Version": "7.3-51.4",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "a489ad42493fa9bfac8f3e3a3bc7d4ca"
    },
    "Matrix": {
      "Package": "Matrix",
      "Version": "1.2-17",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "a60009c832f495604f54399f21d89cc8"
    },
    "boot": {
      "Package": "boot",
      "Version": "1.3-23",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "7acd7cd8da575cbe664ae2301d127007"
    },
    "class": {
      "Package": "class",
      "Version": "7.3-15",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "4878ea322bdd292b96affa71fd4332bd"
    },
    "cluster": {
      "Package": "cluster",
      "Version": "2.1.0",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "d683a0521c6448259522eb63627c3807"
    },
    "codetools": {
      "Package": "codetools",
      "Version": "0.2-16",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "3bf2a0853d7faf7f865d0b877946e452"
    },
    "foreign": {
      "Package": "foreign",
      "Version": "0.8-71",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "0a3e9cb4f946d20e3ce0eae1ae17c746"
    },
    "lattice": {
      "Package": "lattice",
      "Version": "0.20-38",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "6f7aa123115b33bc9ad1ca800874b1de"
    },
    "mgcv": {
      "Package": "mgcv",
      "Version": "1.8-28",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "6ac256e4fdcf33c11f9ed987927cc614"
    },
    "nlme": {
      "Package": "nlme",
      "Version": "3.1-140",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "ad3b02c6f154deac3584133e6e015cd3"
    },
    "nnet": {
      "Package": "nnet",
      "Version": "7.3-12",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "2a92e339b2aeab90147147e2cf7c8310"
    },
    "renv": {
      "Package": "renv",
      "Version": "0.8.2",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "0b45cd1dfa978e60de94abd7084079a0"
    },
    "rpart": {
      "Package": "rpart",
      "Version": "4.1-15",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "0d0e7f0e3c22944bb9bf67d0d5636e50"
    },
    "spatial": {
      "Package": "spatial",
      "Version": "7.3-11",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "0037d3d19ad8a2bf80abb4d7507b746d"
    },
    "survival": {
      "Package": "survival",
      "Version": "2.44-1.1",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "20bdedb74dc2c866f51e4ec9133480b1"
    }
  }
}
```

**Comment by kevinushey** [comment]
I'm surprised that `BiocVersion` isn't included in the lockfile. This is what `renv` (through `BiocManager`) uses to determine the most appropriate repository URLs.

Is the BiocVersion package installed in your project? If not, do you know why? Normally attempts to install Bioconductor packages should also force an install of the BiocVersion package, if I understand correctly.

**Comment by kevinushey** [comment]
You can see here that we just use whatever repositories that BiocManager tells us to:

https://github.com/rstudio/renv/blob/a8395cc9ee16400aef046a559577420d9eace1cd/R/bioconductor.R#L32-L62

And that in turn should be powered by the version of BiocVersion installed on the system.

**Comment by PieGes** [comment]
It seems that when settings `options(repos = BiocManager::repositories(version = "3.9"))` and then installing packages through `renv::install` BiocVersion is not installed. 
I tried to tried to install with `BiocManager::install`, BiocVersion is then indeed recorded in the lockfile but the restore still fail. 

```
> renv::restore()
The following package(s) will be updated:

# Bioconductor =======================
- Biobase        [* -> 2.44.0]
- BiocGenerics   [* -> 0.30.0]
- BiocVersion    [* -> 3.9.0]

# CRAN ===============================
- renv           [0.8.2 -> 0.8.3]
- BiocManager    [* -> 1.30.9]

Do you want to proceed? [y/N]: y
* Querying repositories for available source packages ... Done!
Retrieving 'https://cloud.r-project.org/src/contrib/renv_0.8.3.tar.gz' ...
	OK [file is up to date]
* Querying repositories for available source packages ... Done!
Retrieving 'https://bioconductor.org/packages/3.10/bioc/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...
Retrieving 'https://bioconductor.org/packages/3.10/data/annotation/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...
Retrieving 'https://bioconductor.org/packages/3.10/data/experiment/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...
Retrieving 'https://bioconductor.org/packages/3.10/workflows/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...
Retrieving 'https://cloud.r-project.org/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...
Erreur : failed to retrieve package 'BiocVersion'
De plus : Warning messages:
1: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
2: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
3: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
4: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
5: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
6: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
7: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
8: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
9: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
10: In downloader(url, destfile, type, request, headers) :
  curl: (22) The requested URL returned error: 404 Not Found
Traceback (most recent calls first):
9: renv::restore()
8: renv_restore_run_actions(project, diff, current, lockfile)
7: renv_retrieve(packages)
6: handler(package, renv_retrieve_impl(package))
5: renv_retrieve_impl(package)
4: renv_retrieve_bioconductor(record)
3: renv_retrieve_repos(record)
2: stopf("failed to retrieve package '%s'", record$Package)
1: stop(sprintf(fmt, ...), call. = call.)
```

The lockfile:

```
{
  "R": {
    "Version": "3.6.0",
    "Repositories": [
      {
        "Name": "CRAN",
        "URL": "https://cloud.r-project.org"
      }
    ]
  },
  "Packages": {
    "Biobase": {
      "Package": "Biobase",
      "Version": "2.44.0",
      "Source": "Bioconductor",
      "Hash": "4d603778097e0917bcd309eff2f314f9"
    },
    "BiocGenerics": {
      "Package": "BiocGenerics",
      "Version": "0.30.0",
      "Source": "Bioconductor",
      "Hash": "0e4a1c136ce54159dbc4743379861d92"
    },
    "BiocManager": {
      "Package": "BiocManager",
      "Version": "1.30.9",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "ae39a770c7be4816960bde0c25626b20"
    },
    "BiocVersion": {
      "Package": "BiocVersion",
      "Version": "3.9.0",
      "Source": "Bioconductor",
      "Hash": "7ecece6a1cbb29b32109ec9443e8abad"
    },
    "KernSmooth": {
      "Package": "KernSmooth",
      "Version": "2.23-15",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "1abaf80e4ce51c982446637b6cf236bb"
    },
    "MASS": {
      "Package": "MASS",
      "Version": "7.3-51.4",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "a489ad42493fa9bfac8f3e3a3bc7d4ca"
    },
    "Matrix": {
      "Package": "Matrix",
      "Version": "1.2-17",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "a60009c832f495604f54399f21d89cc8"
    },
    "boot": {
      "Package": "boot",
      "Version": "1.3-23",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "7acd7cd8da575cbe664ae2301d127007"
    },
    "class": {
      "Package": "class",
      "Version": "7.3-15",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "4878ea322bdd292b96affa71fd4332bd"
    },
    "cluster": {
      "Package": "cluster",
      "Version": "2.1.0",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "d683a0521c6448259522eb63627c3807"
    },
    "codetools": {
      "Package": "codetools",
      "Version": "0.2-16",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "3bf2a0853d7faf7f865d0b877946e452"
    },
    "foreign": {
      "Package": "foreign",
      "Version": "0.8-71",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "0a3e9cb4f946d20e3ce0eae1ae17c746"
    },
    "lattice": {
      "Package": "lattice",
      "Version": "0.20-38",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "6f7aa123115b33bc9ad1ca800874b1de"
    },
    "mgcv": {
      "Package": "mgcv",
      "Version": "1.8-28",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "6ac256e4fdcf33c11f9ed987927cc614"
    },
    "nlme": {
      "Package": "nlme",
      "Version": "3.1-140",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "ad3b02c6f154deac3584133e6e015cd3"
    },
    "nnet": {
      "Package": "nnet",
      "Version": "7.3-12",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "2a92e339b2aeab90147147e2cf7c8310"
    },
    "renv": {
      "Package": "renv",
      "Version": "0.8.3",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "4ee3e91b679fbd7102c1fd51f9a4bc59"
    },
    "rpart": {
      "Package": "rpart",
      "Version": "4.1-15",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "0d0e7f0e3c22944bb9bf67d0d5636e50"
    },
    "spatial": {
      "Package": "spatial",
      "Version": "7.3-11",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "0037d3d19ad8a2bf80abb4d7507b746d"
    },
    "survival": {
      "Package": "survival",
      "Version": "2.44-1.1",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "20bdedb74dc2c866f51e4ec9133480b1"
    }
  }
}
```

**Comment by kevinushey** [comment]
> Retrieving 'https://bioconductor.org/packages/3.10/bioc/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...
> Retrieving 'https://bioconductor.org/packages/3.10/data/annotation/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...
> Retrieving 'https://bioconductor.org/packages/3.10/data/experiment/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...
> Retrieving 'https://bioconductor.org/packages/3.10/workflows/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...
> Retrieving 'https://cloud.r-project.org/src/contrib/Archive/BiocVersion/BiocVersion_3.9.0.tar.gz' ...

It looks like `renv` is still using Bioconductor 3.10 during restore, hence it fails -- even though BiocVersion 3.9 is installed. :-/

I've just pushed a commit to `renv` -- now, we explicitly track the version of Bioconductor used in the lockfile, and pass that along to `BiocManager` / `BiocInstaller` when requesting the appropriate repositories for use.

Can you try snapshotting + restoring with the development version of `renv` and let me know if things are better there?

**Comment by PieGes** [comment]
Thanks! I was able to restore with Bioconductor 3.9 with renv 0.8.3-24. However we need to explicitly install BiocVersion with `BiocManager::install(version="3.9")` to get the correct version recorded in the lockfile (setting the repos to `BiocManager::repositories(version="3.9")` is insufficient). 

I also faced a strange behaviour when doing `BiocManager::install(version="3.9")`, renv is downgraded to 0.8.3:

```
* Project '/bioinfo/users/gestraud/R/test/bioc' loaded. [renv 0.8.3-24]
> renv::install("BiocManager")
Retrieving 'https://cloud.r-project.org/src/contrib/BiocManager_1.30.10.tar.gz' ...
	OK [file is up to date]
Installing BiocManager [1.30.10] ...
	OK (linked cache)
> BiocManager::install(version="3.9")
Downgrade 2 packages to Bioconductor version '3.9'? [y/n]: y
Bioconductor version 3.9 (BiocManager 1.30.10), R 3.6.0 (2019-04-26)
Installing package(s) 'BiocVersion'
essai de l'URL 'https://bioconductor.org/packages/3.9/bioc/src/contrib/BiocVersion_3.9.0.tar.gz'
Content type 'application/x-gzip' length 984 bytes
==================================================
downloaded 984 bytes

* installing *source* package ‘BiocVersion’ ...
** using staged installation
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (BiocVersion)

The downloaded source packages are in
	‘/tmp/RtmpmGcFln/downloaded_packages’
Installing package(s) 'renv'
essai de l'URL 'https://cloud.r-project.org/src/contrib/renv_0.8.3.tar.gz'
Content type 'application/x-gzip' length 213672 bytes (208 KB)
==================================================
downloaded 208 KB

* installing *source* package ‘renv’ ...
** package ‘renv’ correctement décompressé et sommes MD5 vérifiées
** using staged installation
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
Warning in read.dcf(file.path(p, "DESCRIPTION"), c("Package", "Version")) :
  impossible d'ouvrir le fichier compressé '/bioinfo/users/gestraud/R/test/bioc/renv/library/R-3.6/x86_64-pc-linux-gnu/renv/DESCRIPTION', cause probable : 'Aucun fichier ou dossier de ce type'
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (renv)

The downloaded source packages are in
	‘/tmp/RtmpmGcFln/downloaded_packages’
```
To get around this I did a snapshot before installation of BiocVersion and a restore right after.

**Comment by kevinushey** [comment]
It looks like `BiocManager` itself is forcing the downgrade, which is pretty odd. Perhaps this gives me more motivation to support `renv::install("bioc::package")` out of the box...

**Comment by kevinushey** [comment]
With the development version of `renv`, you can now install packages from Bioconductor with (for example):

```
renv::install("bioc::limma")
```

In addition, `renv` now also records the Bioconductor version in the lockfile as appropriate; e.g.

```
{
  "R": {
    "Version": "3.5.3",
    "Repositories": [
      {
        "Name": "CRAN",
        "URL": "https://cran.rstudio.com"
      }
    ]
  },
  "Bioconductor": {
    "Version": "3.8"
  },
  "Packages": {
    "BiocManager": {
      "Package": "BiocManager",
      "Version": "1.30.10",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "b07155a0faf734d390fa699cf4403be2"
    },
    "limma": {
      "Package": "limma",
      "Version": "3.38.3",
      "Source": "Bioconductor",
      "Hash": "54efab82b469458f8b17df3000ba2388"
    },
    "renv": {
      "Package": "renv",
      "Version": "0.8.3-33",
      "Source": "Repository",
      "Repository": "CRAN",
      "Hash": "3edbc179660e070087b04845a0c82b15"
    }
  }
}
```

That stored version will be supplied to `BiocManager` as appropriate when determining the appropriate repositories to use.

To wit, I think all the Bioconductor issues outlined here should now be solved -- please let me know if that does not appear to be the case or if you run into other issues.

**Comment by PieGes** [comment]
Many thanks! I'll give a try to this new feature and come back to you if I face something unexpected.

**Comment by krzyslom** [comment]
Hi, I've just stumbled upon a problem that is related to yesterday's feature.

I'm running rocker/verse:`3.5.2` Docker image. Calling `remotes::install_bioc("Rgraphviz")` installs said package but also installs the BiocGenerics one. When I run `renv::snapshot()` it correctly recognizes the version of Bioconductor as `3.8` since it is the latest one for `R-3.5`. However both of the packages were installed for very different Bioconductor versions. The following setup is the result:

Rgraphviz: `2.30.0` from Bioconductor `3.10`,
BiocGenerics: `0.26.0`. from Bioconductor `3.7`

Hence the `renv::restore()` call results in a failure as the mentioned package versions are looked for in Bioconductor `3.8`. At this point it does not matter if I have installed `BiocManager` and `BiocVersion`. Looks like such behavior for `remotes` is implemented on purpose - https://github.com/r-lib/remotes/issues/361#issuecomment-517678616.

The newest feature `renv::install("bioc::Rgraphviz")` (as well as `BiocManager::install`) solves the problem as for R `3.5.2` both packages are installed from Bioconductor `3.8`.

Rgraphviz: `2.26.0` from Bioconductor `3.8`,
BiocGenerics: `0.28.0`. from Bioconductor `3.8`

Strangely for `remotes::install_bioc` the newer version of one package installs the older version of the second one.

The question is should the `renv` package functionality be compatible with the `remotes`? I see that the topic is mentioned in #211. However I'd imagine that the best practice is to always have the Bioconductor packages installed from a single release only. Maybe some kind of warning/error message in `renv::snapshot()` would be sufficient? For example "It seems like the packages do not come from the same Bioconductor release. Please reinstall them with `renv::install` before taking a snapshot."

To sum it up, it is possible for popular solution with `remotes:install_bioc` to fail when working with `renv`. It is better to stick with `renv::install(bioc::<package>)` or `BiocManager::install(<package>)`.

**Comment by kevinushey** [comment]
Thanks for the investigation. It looks like this is indeed the documented behavior for `remotes::install_bioc()`: it installs development packages from the Bioconductor git repository. I'm not sure if the cross-version installation behavior is expected, though.

`renv::install("bioc::<pkg>")` instead tries to install from the "active" Bioconductor release, and coordinates with the `BiocManager` package to determine what repositories should be used in that case.

> However I'd imagine that the best practice is to always have the Bioconductor packages installed from a single release only. Maybe some kind of warning/error message in renv::snapshot() would be sufficient? For example "It seems like the packages do not come from the same Bioconductor release. Please reinstall them with renv::install before taking a snapshot."

This seems reasonable. It looks like Bioconductor packages record the release they're associated with in the DESCRIPTION file; e.g. I see for Biobase in my R 3.5 installation:

```
< ... >
biocViews: Infrastructure
git_url: https://git.bioconductor.org/packages/Biobase
git_branch: RELEASE_3_8
git_last_commit: 3e5bd46
git_last_commit_date: 2018-10-30
< ... >
```

Presumedly, that release tag should normally match with the Bioconductor version (hopefully, at least)

**Comment by krzyslom** [comment]
Thanks for creating this wonderful package.

>This seems reasonable. It looks like Bioconductor packages record the release they're associated with in the DESCRIPTION file; e.g. I see for Biobase in my R 3.5 installation:

Similar lines could be found in `3.8` release of BiocGenerics. Unfortunately it looks like for the `3.7` release the `git_*` lines are absent. I wonder if it is a standard from the `3.8` version up. I can't find any details on that.

**Comment by kevinushey** [comment]
I've now implemented a strategy for handling this -- `renv` should now warn if it detects that one or more of the Bioconductor packages used in a project appear to be from a separate Bioconductor release.
